### PR TITLE
Record: SP4096 + Depth Recurrence + Parallel Residuals + MuonEq-R + QK-Gain 5.0 — val_bpb 1.0897 (3-seed mean)

### PR DESCRIPTION
**Pull request by aryanbhosale**
## Record: SP4096 + Depth Recurrence + Parallel Residuals + MuonEq-R + QK-Gain 5.0

**val_bpb = 1.0897** (3-seed mean, std 0.0003) | **~15.99 MB** | 8×H100 SXM

### 3-Seed Results

| Seed | **Sliding BPB** | Artifact |
|------|-----------------|----------|
| 42   | **1.0894**      | 15,999,533 |
| 314  | **1.0898**      | 15,992,752 |
| 999  | **1.0899**      | 15,988,473 |
| **Mean** | **1.0897** | |

Merged SOTA (PR #1019): **1.1147 BPB**. Delta: **−0.0250 BPB**.

### Key Techniques

1. **4096-Vocab + MLP 4x + WD 0.090** — PR #1218 @clarkkev, PR #1285 @dexhunter
2. **Depth Recurrence (layers 4,5)** — virtual 13-layer from 11 physical. PR #1204 @msisovic, PR #1260 @dexhunter
3. **Parallel Residuals (from layer 7)** — separate attn/MLP lanes. PR #1204 @msisovic, PR #1289 @MatoTeziTanka
4. **MuonEq-R** — arXiv:2603.28254. PR #1260 @dexhunter
5. **QK-Gain 5.0** — PR #1217 @bigbag
6. **Full GPTQ int6 + Brotli + LZMA Compressed Wrapper** (~24KB code)

### Compliance

- **No TTT** — no weight updates during evaluation
- **No SLOT** — no eval-time delta optimization
- **No n-gram cache** — no eval-time statistics
- **No eval-time adaptation of any kind** — model weights completely frozen
- GPTQ calibration within training budget
- Standard autoregressive sliding-window eval (stride=64)
- All four conditions from Issue #1017 satisfied

### Reproduction

```bash
pip install brotli
MATCHED_FINEWEB_REPO_ID=kevclark/parameter-golf python3 data/cached_challenge_fineweb.py --variant sp4096 --skip-manifest
SEED=42 RECUR_LAYERS=4,5 RECUR_START_STEP=3000 PARALLEL_START_LAYER=7 \
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

### Credits

PR #1218 @clarkkev, PR #1285 @dexhunter, PR #1204 @msisovic, PR #1289 @MatoTeziTanka, PR #1260 @dexhunter, PR #1019 @abaybektursun, PR #1287 @dentity007, PR #1217 @bigbag, PR #493 @parinzee

### CODE REVIEW

**Comment by cocohearts** [comment]
@aryanbhosale maintainer repro note: we are doing a pass over merged leaderboard rows and I cannot currently reproduce this record closely enough.

What I ran:
- Exact submitted record archive for PR #1334.
- Same seed `42`, same visible hparameter block, same exact ISO FlashAttention wheel we are using for the audit.
- I noticed our first repro had staged `train_shards: 143` while your submitted log says `train_shards: 80`, so I reran with the staged dataset forced to `80` shards.

Current result:
- Submitted seed `42`: `final_int6_sliding_window val_bpb=1.08938974`, stop step `5453`.
- Targeted repro seed `42` with `train_shards: 80`: `final_int6_sliding_window val_bpb=1.09250239`, stop step `5247`.
- Initial 3-seed repro mean was `1.09215243` vs submitted mean `1.08971631`.

The visible hparams and the train-shard count now match, but the repro is still slower after recurrence / stops earlier and enters GPTQ with a larger int6 artifact (`unpruned=16.03MB`, requiring pruning) whereas the submitted seed42 log already fit at about `16.00MB`.

Could you reply here with any missing exact setup details that might explain this? In particular: exact runtime/container, Python/PyTorch/CUDA/FlashAttention versions, dataset snapshot/prep command, or any env vars not captured in the submitted log. If there is an updated record/log bundle that reproduces the submitted numbers, please point us to it or push it so we can rerun.

**Comment by aryanbhosale** [comment]
> @aryanbhosale maintainer repro note: we are doing a pass over merged leaderboard rows and I cannot currently reproduce this record closely enough.
> 
> What I ran:
> 
> * Exact submitted record archive for PR [Record: SP4096 + Depth Recurrence + Parallel Residuals + MuonEq-R + QK-Gain 5.0 — val_bpb 1.0897 (3-seed mean) #1334](https://github.com/openai/parameter-golf/pull/1334).
> * Same seed `42`, same visible hparameter block, same exact ISO FlashAttention wheel we are using for the audit.
> * I noticed our first repro had staged `train_shards: 143` while your submitted log says `train_shards: 80`, so I reran with the staged dataset forced to `80` shards.
> 
> Current result:
> 
> * Submitted seed `42`: `final_int6_sliding_window val_bpb=1.08938974`, stop step `5453`.
> * Targeted repro seed `42` with `train_shards: 80`: `final_int6_sliding_window val_bpb=1.09250239`, stop step `5247`.
> * Initial 3-seed repro mean was `1.09215243` vs submitted mean `1.08971631`.
> 
> The visible hparams and the train-shard count now match, but the repro is still slower after recurrence / stops earlier and enters GPTQ with a larger int6 artifact (`unpruned=16.03MB`, requiring pruning) whereas the submitted seed42 log already fit at about `16.00MB`.
> 
> Could you reply here with any missing exact setup details that might explain this? In particular: exact runtime/container, Python/PyTorch/CUDA/FlashAttention versions, dataset snapshot/prep command, or any env vars not captured in the submitted log. If there is an updated record/log bundle that reproduces the submitted numbers, please point us to it or push it so we can rerun.

Hey @cocohearts, thanks for digging into this — and sorry the bundle didn't carry enough to repro cleanly the first time.

I think the gap is kernel-level, not config. Here's why I'm fairly sure:

The env actually came from PR #1019 (@abaybektursun's record, which #1334 is built on top of). Its `submission.json` pins the stack explicitly:

```
pytorch_version: "2.9.1+cu128"
cuda_version: "12.8"
flash_attn_version: "2.8.3 (FA3 Hopper kernels)"
hardware: "8xH100 80GB SXM"
```

I ran #1334 on the same pod with that same install untouched — my `submission.json` only re-stated `pytorch_version` and dropped the other two fields, which is on me. So it's **flash-attn 2.8.3 (FA3 Hopper) + torch 2.9.1+cu128 + CUDA 12.8**, not FA2, not torch 2.11.

The reason I'm pretty confident this is environmental: I ran three different SP4096 architectures on April 3–4 against that same install, and seed-42 stopped at almost exactly the same step every time:

| Submission | seed-42 stop | train_time | Log |
|---|---|---|---|
| `sp4096-depth-recurrence-muoneqr` (#1334's lineage) | 5453 | 590 026 ms | [train_seed42.log](https://github.com/aryanbhosale/parameter-golf/blob/submission/sp4096-depth-recurrence-muoneqr/records/track_10min_16mb/2026-04-03_SP4096_DepthRecurrence_MuonEqR_GPTQ/train_seed42.log) |
| `sp4096-causal-slot` (different head) | 5454 | 590 101 ms | [train_seed42.log](https://github.com/aryanbhosale/parameter-golf/blob/submission/sp4096-causal-slot/records/track_10min_16mb/2026-04-04_SP4096_DepthRecurrence_ParallelResid_CausalSLOT/train_seed42.log) |
| `sp4096-v6-ttt` (#1334 trunk + TTT) | 5440 | 590 016 ms | [train_seed42.log](https://github.com/aryanbhosale/parameter-golf/blob/submission/sp4096-v6-ttt/records/track_10min_16mb/2026-04-04_SP4096_DepthRecurrence_ParallelResid_MuonEqR_TTT/train_seed42.log) |

Same `model_params: 34401372`, same `train_shards: 80`, same wallclock cap — three architecturally different runs all landed in a 14-step band at 5440–5454. Your audit at 5247 is ~200 below the whole cluster (~3.8% slower per step). A config or data drift would shift one of those, not pull all three down by the same amount, so I really do think it's the kernel.

If you can pin the audit container to:

- `torch==2.9.1+cu128` (the cu128 PyPI wheel, not cu130)
- FA3 source-built from `flash-attn==2.8.3`, `hopper/` subdir, against torch 2.9.1+cu128 + CUDA 12.8 toolkit
- inductor cache pre-warmed (the wrapper has three `torch.compile(fullgraph=True, dynamic=False)` sites, plus a recompile when `RECUR_START_STEP=3000` flips the model to virtual-13L — on a cold cache that recompile gets paid inside the 590 s budget; my submitted run had it warm from the dry-run)

…I'd expect seed-42 to land back in the 5440–5454 band and the int6 artifact to fit at 16.00 MB without prune cycles.

One thing worth flagging since you're reading the log directly: `train_seed42.log` tails out with `final_int6_ttt val_bpb=1.33603200`. That's leftover instrumentation — `ttt_enabled: True` got left in the wrapper's hparams block, so the script ran a TTT eval after the scored sliding-window eval. It's **not** the scored number for this Track-A submission; the scored value is `final_int6_sliding_window val_bpb=1.08938974` with weights frozen. If anything in the audit pipeline tail-greps the log, that line could mislead.

The rental pod's gone so I can't pull its container hash, but I do still have the FA3 wheel artifact I built. Happy to send it directly, or push a `repro/` subdir under the #1334 branch on my fork (https://github.com/aryanbhosale/parameter-golf/tree/submission/sp4096-no-slot-v4) with a fresh log on a freshly-provisioned 8×H100 against the pinned stack — whichever's more useful for the audit. The wheel's faster; let me know.
